File: src/web/src/lib/slash-commands.ts
import type { Bootstrap } from './api/types'

export interface SlashCommand {
  value: string
  description: string
  label?: string
  keywords?: string
  kind?: 'command' | 'skill' | 'model' | 'mode'
}

export const baseSlashCommands: SlashCommand[] = [
  { value: '/model', label: '模型', description: '选择下一轮使用的模型', keywords: '模型 model' },
  { value: '/thinking', label: '推理强度', description: '选择此任务下一轮的推理档位', keywords: 'reasoning effort' },
  { value: '/tasks', label: '计划记录', description: '在运行记录中查看计划历史', keywords: 'plan 计划 步骤' },
  { value: '/mode', label: '审批模式', description: '选择工具操作的确认方式', keywords: 'approval' },
  { value: '/copy', label: '复制回答', description: '将最新回复复制到剪贴板' },
  { value: '/new', label: '新建任务', description: '开始一段新对话', keywords: 'session task' },
  { value: '/retry', description: '重试上次任务', keywords: 'again run' },
  { value: '/agents', description: '查看和协调子 Agent', keywords: 'subagent worker' },
  { value: '/checkpoints', label: '分支与历史', description: '查看检查点，从历史创建分支', keywords: 'rewind fork' },
  { value: '/transcript', description: '搜索结构化 transcript', keywords: 'audit history' },
  { value: '/dequeue', description: '撤回尚未执行的排队输入', keywords: 'queue restore' },
  { value: '/clear', description: '清空当前显示', keywords: 'timeline' },
  {
    value: '/plan ',
    description: '先生成计划，确认后再执行任务',
    keywords: 'planning collaboration',
    kind: 'mode',
  },
  { value: '/context', description: '查看上下文', keywords: 'tokens' },
  { value: '/context sources', description: '查看当前会话上下文来源' },
  { value: '/instructions', description: '查看稳定 prompt、动态上下文与来源' },
  { value: '/compact ', description: '压缩上下文', keywords: 'focus' },
  { value: '/memory ', description: '管理记忆', keywords: 'remember forget list' },
  { value: '/tools', label: '工具', description: '查看当前可用能力' },
  { value: '/skills', label: '技能', description: '查看可用的领域流程' },
  { value: '/skill unload ', description: '卸载当前会话 Skill' },
  { value: '/resource ', description: '激活 MCP resource' },
  { value: '/resource refresh ', description: '刷新当前会话 MCP resource 快照' },
  { value: '/resource unload ', description: '卸载 MCP resource' },
  { value: '/clarification cancel', description: '取消待回答澄清' },
  { value: '/mcp', description: '查看 MCP 连接' },
  { value: '/prompts', description: '查看 MCP prompt 模板' },
  { value: '/prompt ', description: '渲染并运行 MCP prompt 模板' },
  { value: '/hooks', description: '查看 hooks 与触发统计' },
  { value: '/help', description: '查看命令帮助' },
]

/** Display labels, help text, and inserted commands have separate roles. */
export function buildSlashCommands(
  bootstrap?: Pick<Bootstrap, 'availableModels' | 'activeModel' | 'skills'> | null,
): SlashCommand[] {
  return [
    ...baseSlashCommands,
    ...(bootstrap?.availableModels ?? []).map((model): SlashCommand => ({
      value: `/model ${model}`,
      label: model,
      description: model === bootstrap?.activeModel ? '当前模型' : '切换模型',
      keywords: 'model',
      kind: 'model',
    })),
    { value: '/mode manual', label: '每次确认', description: '操作前请求确认', keywords: 'approval', kind: 'mode' },
    { value: '/mode accept_edits', label: '自动接受文件修改', description: '其他操作仍按策略审批', keywords: 'approval', kind: 'mode' },
    { value: '/mode auto', label: '自动执行', description: '按自动审批策略执行', keywords: 'approval', kind: 'mode' },
    ...(bootstrap?.skills ?? []).map((skill): SlashCommand => ({
      value: `/skill:${skill.name} `,
      label: skill.name,
      description: skill.description || '运行技能',
      keywords: 'skill',
      kind: 'skill',
    })),
  ]
}

export interface PromptInvocation {
  reference: string
  arguments: Record<string, string>
}

function splitCommandLine(value: string): string[] {
  const words: string[] = []
  let word = ''
  let quote: 'single' | 'double' | null = null
  let escaped = false
  let started = false

  for (const character of value.trim()) {
    if (escaped) {
      word += character
      escaped = false
      started = true
      continue
    }
    if (character === '\\' && quote !== 'single') {
      escaped = true
      started = true
      continue
    }
    if (character === "'" && quote !== 'double') {
      quote = quote === 'single' ? null : 'single'
      started = true
      continue
    }
    if (character === '"' && quote !== 'single') {
      quote = quote === 'double' ? null : 'double'
      started = true
      continue
    }
    if (/\s/.test(character) && quote === null) {
      if (started) words.push(word)
      word = ''
      started = false
      continue
    }
    word += character
    started = true
  }
  if (quote !== null) throw new Error('prompt 参数的引号未闭合')
  if (escaped) word += '\\'
  if (started) words.push(word)
  return words
}

export function parsePromptInvocation(value: string): PromptInvocation {
  const words = splitCommandLine(value)
  if (words[0]?.toLowerCase() !== '/prompt' || !words[1]) {
    throw new Error('用法：/prompt <server:name> [key=value ...]')
  }
  const arguments_: Record<string, string> = {}
  for (const item of words.slice(2)) {
    const separator = item.indexOf('=')
    if (separator <= 0) throw new Error(`无效 prompt 参数：${item}；请使用 key=value`)
    arguments_[item.slice(0, separator)] = item.slice(separator + 1)
  }
  return { reference: words[1], arguments: arguments_ }
}

export function slashQuery(value: string): string | null {
  return /^\/\S*$/.test(value) ? value.slice(1).toLowerCase() : null
}

export function filterSlashCommands(commands: SlashCommand[], value: string): SlashCommand[] {
  const argumentQuery = value.match(/^\/(model|mode)\s+(.*)$/i)
  if (argumentQuery) {
    const prefix = `/${argumentQuery[1].toLowerCase()} `
    const query = argumentQuery[2].trim().toLowerCase()
    return commands.filter((item) => item.value.toLowerCase().startsWith(prefix)
      && item.value.slice(prefix.length).toLowerCase().includes(query))
      .sort((a, b) => Number(b.value.slice(prefix.length).toLowerCase() === query)
        - Number(a.value.slice(prefix.length).toLowerCase() === query))
  }
  const query = slashQuery(value)
  if (query == null) return []
  if (!query) return commands.filter((item) => !/^\/(model|mode)\s/.test(item.value))
  return commands.filter((item) => {
    const haystack = `${item.value.slice(1)} ${item.label ?? ''} ${item.description} ${item.keywords ?? ''}`.toLowerCase()
    return haystack.includes(query)
  })
}

File: src/lumen/ui/slash_commands.py
"""Slash command registry — the single source of truth for TUI slash commands.

Every consumer derives from this module instead of keeping its own hand-written
list:

* dispatch (``app._handle_command``) looks up the entry and calls its handler;
* ``/help`` renders the grouped, per-command reference via :func:`render_help`;
* the completion dropdown lists :func:`iter_visible` entries plus their
  ``completion_rows``;
* the run-time safety gate (``command_gate``) maps ``while_running`` to a
  ``CommandPolicy``;
* the command palette (``commands.py``) pulls help text from ``in_palette``
  entries.

``/skill:<name>`` keeps its special prefix-parsing branch in the dispatcher
(the name is dynamic); the registry only carries its help/completion text via
the ``skill:`` pseudo-entry.
"""

from __future__ import annotations

from collections.abc import Iterator
from dataclasses import dataclass

from lumen.branding import FRAMEWORK_NAME

# Run-safety policies, mirroring command_gate.CommandPolicy values (kept as
# plain strings so this module stays import-free of the gate).
ALLOW = "allow"
BLOCK = "block"
QUEUE = "queue"
CANCEL_THEN_RUN = "cancel_then_run"

# Help categories in display order.
CATEGORY_SESSION = "Session"
CATEGORY_MODEL = "Model"
CATEGORY_CONTEXT = "Context"
CATEGORY_MCP = "MCP"
CATEGORY_MEMORY = "Memory"
CATEGORY_OTHER = "Other"
CATEGORY_ORDER: tuple[str, ...] = (
    CATEGORY_SESSION,
    CATEGORY_MODEL,
    CATEGORY_CONTEXT,
    CATEGORY_MCP,
    CATEGORY_MEMORY,
    CATEGORY_OTHER,
)


@dataclass(frozen=True)
class SlashCommand:
    """One slash command.

    ``name`` is the command without the leading slash (``"model"``); the
    ``"skill:"`` pseudo-entry documents the dynamic ``/skill:<name>`` form.
    ``handler`` is the ``LumenApp`` method that runs the command
    (``_cmd_xxx(parts, raw)``); empty for the dispatcher-handled ``skill:``
    branch. ``while_running`` is the run-safety policy consumed by
    ``command_gate``. ``completion_rows`` are extra subcommand rows shown in
    the completion dropdown as ``/<name> <suffix>``.
    """

    name: str
    usage: str
    description: str
    category: str
    handler: str
    while_running: str = ALLOW
    aliases: tuple[str, ...] = ()
    in_palette: bool = False
    hidden: bool = False
    completion_rows: tuple[tuple[str, str], ...] = ()


_COMMANDS: tuple[SlashCommand, ...] = (
    # --- session lifecycle -------------------------------------------------
    SlashCommand(
        name="new",
        usage="/new",
        description="Start a fresh session",
        category=CATEGORY_SESSION,
        handler="_cmd_new",
        while_running=BLOCK,
        in_palette=True,
    ),
    SlashCommand(
        name="sessions",
        usage="/sessions",
        description="List past sessions",
        category=CATEGORY_SESSION,
        handler="_cmd_sessions",
        in_palette=True,
    ),
    SlashCommand(
        name="agents",
        usage="/agents [interrupt <id>]",
        description="Inspect or interrupt persistent Agents",
        category=CATEGORY_SESSION,
        handler="_cmd_children",
        in_palette=False,
        hidden=True,
        completion_rows=(("interrupt", "Interrupt one Agent"),),
    ),
    SlashCommand(
        name="children",
        usage="/children [cancel <id>]",
        description="Inspect or cancel persistent child tasks",
        category=CATEGORY_SESSION,
        handler="_cmd_children",
        in_palette=True,
        completion_rows=(("cancel", "Cancel one child task"),),
    ),
    SlashCommand(
        name="checkpoints",
        usage="/checkpoints",
        description="Browse receipts and rewind into a new session branch",
        category=CATEGORY_SESSION,
        handler="_cmd_checkpoints",
        while_running=BLOCK,
        in_palette=True,
    ),
    SlashCommand(
        name="resume",
        usage="/resume <id>",
        description="Resume a session by id",
        category=CATEGORY_SESSION,
        handler="_cmd_resume",
        while_running=BLOCK,
        in_palette=True,
    ),
    SlashCommand(
        name="clear",
        usage="/clear",
        description="Clear the visible timeline · keep session context",
        category=CATEGORY_SESSION,
        handler="_cmd_clear",
        while_running=BLOCK,
        in_palette=True,
    ),
    SlashCommand(
        name="retry",
        usage="/retry",
        description="Re-send the last prompt",
        category=CATEGORY_SESSION,
        handler="_cmd_retry",
        while_running=QUEUE,
        in_palette=True,
    ),
    SlashCommand(
        name="edit",
        usage="/edit",
        description="Edit the last prompt in $EDITOR and resend on a fresh branch",
        category=CATEGORY_SESSION,
        handler="_cmd_edit",
        while_running=BLOCK,
        in_palette=True,
    ),
    # --- model / approval ---------------------------------------------------
    SlashCommand(
        name="thinking",
        usage="/thinking [level]",
        description="Choose reasoning effort for this Session",
        category=CATEGORY_MODEL,
        handler="_cmd_thinking",
        in_palette=True,
    ),
    SlashCommand(
        name="model",
        usage="/model [name]",
        description="Choose a model or switch by name",
        category=CATEGORY_MODEL,
        handler="_cmd_model",
        # The picker is read-only during a run; switching rebuilds the
        # runtime. The gate refines this via classify_model_command.
        in_palette=True,
    ),
    SlashCommand(
        name="mode",
        usage="/mode [manual|accept_edits|plan|auto]",
        description="Choose a permission mode or switch by name",
        category=CATEGORY_MODEL,
        handler="_cmd_mode",
        in_palette=True,
    ),
    SlashCommand(
        name="tasks",
        usage="/tasks",
        description="Show or collapse the latest task plan",
        category=CATEGORY_SESSION,
        handler="_cmd_tasks",
    ),
    SlashCommand(
        name="status",
        usage="/status",
        description="Show workspace, session, modes and UI settings",
        category=CATEGORY_MODEL,
        handler="_cmd_status",
        in_palette=True,
    ),
    # --- context -------------------------------------------------------------
    SlashCommand(
        name="context",
        usage="/context [--json|sources]",
        description="Show context budget · zones, blocks, pressure",
        category=CATEGORY_CONTEXT,
        handler="_cmd_context",
    ),
    SlashCommand(
        name="instructions",
        usage="/instructions [--json]",
        description="查看稳定 prompt、动态上下文和来源摘要",
        category=CATEGORY_CONTEXT,
        handler="_cmd_instructions",
    ),
    SlashCommand(
        name="compact",
        usage="/compact [focus]",
        description="Force a compaction (M4)",
        category=CATEGORY_CONTEXT,
        handler="_cmd_compact",
    ),
    SlashCommand(
        name="clarification",
        usage="/clarification cancel",
        description="Cancel the pending clarification",
        category=CATEGORY_CONTEXT,
        handler="_cmd_clarification",
        completion_rows=(("cancel", "Cancel the pending clarification"),),
    ),
    # --- MCP ------------------------------------------------------------------
    SlashCommand(
        name="mcp",
        usage="/mcp",
        description="MCP connections and deferred schema status",
        category=CATEGORY_MCP,
        handler="_cmd_mcp",
    ),
    SlashCommand(
        name="resources",
        usage="/resources",
        description="List MCP resources available for explicit context loading",
        category=CATEGORY_MCP,
        handler="_cmd_resources",
    ),
    SlashCommand(
        name="resource",
        usage="/resource [refresh|unload] <ref>",
        description="Load one MCP resource into the retrieved-context zone",
        category=CATEGORY_MCP,
        handler="_cmd_resource",
        completion_rows=(
            ("refresh", "Refresh one session resource snapshot"),
            ("unload", "Unload one resource from this session"),
        ),
    ),
    SlashCommand(
        name="prompts",
        usage="/prompts",
        description="List MCP prompt templates",
        category=CATEGORY_MCP,
        handler="_cmd_prompts",
    ),
    SlashCommand(
        name="prompt",
        usage="/prompt <server:name> [key=value ...]",
        description="Render and submit an MCP prompt template",
        category=CATEGORY_MCP,
        handler="_cmd_prompt",
        while_running=QUEUE,
    ),
    # --- memory ---------------------------------------------------------------
    SlashCommand(
        name="memory",
        usage="/memory [list|remember|edit|forget|use|learn|incognito|rebuild]",
        description="Memory: list, remember, edit, forget, use, learn, rebuild",
        category=CATEGORY_MEMORY,
        handler="_cmd_memory",
    ),
    # --- other ------------------------------------------------------------------
    SlashCommand(
        name="theme",
        usage="/theme [lumen-dark|lumen-light]",
        description="List available themes or switch the active UI theme",
        category=CATEGORY_OTHER,
        handler="_cmd_theme",
    ),
    SlashCommand(
        name="transcript",
        usage="/transcript",
        description="Open searchable structured transcript",
        category=CATEGORY_OTHER,
        handler="_cmd_transcript",
        in_palette=True,
    ),
    SlashCommand(
        name="help",
        usage="/help",
        description="Show this command reference",
        category=CATEGORY_OTHER,
        handler="_cmd_help",
    ),
    SlashCommand(
        name="tools",
        usage="/tools",
        description="List model-visible tools",
        category=CATEGORY_OTHER,
        handler="_cmd_tools",
        in_palette=True,
    ),
    SlashCommand(
        name="hooks",
        usage="/hooks",
        description="List configured hooks and statistics",
        category=CATEGORY_OTHER,
        handler="_cmd_hooks",
        in_palette=True,
    ),
    SlashCommand(
        name="copy",
        usage="/copy",
        description="Copy the latest assistant response",
        category=CATEGORY_OTHER,
        handler="_cmd_copy",
        in_palette=True,
    ),
    SlashCommand(
        name="skills",
        usage="/skills",
        description="List available skills",
        category=CATEGORY_OTHER,
        handler="_cmd_skills",
    ),
    SlashCommand(
        name="skill:",
        usage="/skill:<name> [args]",
        description="Manually trigger a skill",
        category=CATEGORY_OTHER,
        handler="",  # dynamic prefix form; dispatched by app._handle_command
        while_running=QUEUE,
    ),
    SlashCommand(
        name="skill",
        usage="/skill unload <name>",
        description="Unload a skill from this session",
        category=CATEGORY_OTHER,
        handler="_cmd_skill",
        completion_rows=(("unload", "Unload one skill from this session"),),
    ),
    SlashCommand(
        name="exit",
        usage="/exit",
        description=f"Exit {FRAMEWORK_NAME}",
        category=CATEGORY_OTHER,
        handler="_cmd_exit",
        while_running=CANCEL_THEN_RUN,
        # /quit stays as a hidden alias for existing users: it dispatches and
        # gates like /exit but is kept out of /help and completions.
        aliases=("quit",),
        in_palette=True,
    ),
)

_BY_NAME: dict[str, SlashCommand] = {}
for _command in _COMMANDS:
    _BY_NAME[_command.name] = _command
    for _alias in _command.aliases:
        _BY_NAME[_alias] = _command


def find_command(name: str) -> SlashCommand | None:
    """Look up a command by name (case-insensitive, leading ``/`` tolerated).

    Aliases resolve to their canonical entry, so ``find_command("quit")``
    returns the ``exit`` entry.
    """

    return _BY_NAME.get(name.lower().lstrip("/"))


def iter_visible() -> Iterator[SlashCommand]:
    """Yield non-hidden commands in registry (help) order."""

    return (command for command in _COMMANDS if not command.hidden)


def render_help() -> str:
    """Render the grouped ``/help`` reference.

    One header per category; each command is ``usage  —  description``,
    aligned within its group. Hidden commands are omitted.
    """

    lines: list[str] = []
    for category in CATEGORY_ORDER:
        entries = [command for command in _COMMANDS if command.category == category and not command.hidden]
        if not entries:
            continue
        if lines:
            lines.append("")
        lines.append(f"{category}:")
        width = max(len(command.usage) for command in entries)
        for command in entries:
            lines.append(f"  {command.usage.ljust(width)}  —  {command.description}")
    return "\n".join(lines)


__all__ = [
    "ALLOW",
    "BLOCK",
    "CANCEL_THEN_RUN",
    "CATEGORY_ORDER",
    "QUEUE",
    "SlashCommand",
    "find_command",
    "iter_visible",
    "render_help",
]
